File: src/test/java/com/enonic/app/oidcidprovider/lib/GroupSyncTest.java
package com.enonic.app.oidcidprovider.lib;

import org.junit.jupiter.api.Test;

import com.enonic.xp.testing.ScriptTestSupport;

public class GroupSyncTest
    extends ScriptTestSupport
{
    private void run( final String function )
    {
        runFunction( "/lib/groupSync-test.js", function );
    }

    @Test
    public void testResolve_disabledFeature_returnsEmpty()
    {
        run( "testResolve_disabledFeature_returnsEmpty" );
    }

    @Test
    public void testResolve_missingPath_returnsEmpty()
    {
        run( "testResolve_missingPath_returnsEmpty" );
    }

    @Test
    public void testResolve_topLevelArray_returnsMappedKeys()
    {
        run( "testResolve_topLevelArray_returnsMappedKeys" );
    }

    @Test
    public void testResolve_singleString_treatedAsSingleton()
    {
        run( "testResolve_singleString_treatedAsSingleton" );
    }

    @Test
    public void testResolve_nestedDottedPath()
    {
        run( "testResolve_nestedDottedPath" );
    }

    @Test
    public void testResolve_uriClaim_usedAsLiteralKey()
    {
        run( "testResolve_uriClaim_usedAsLiteralKey" );
    }

    @Test
    public void testResolve_urnClaim_usedAsLiteralKey()
    {
        run( "testResolve_urnClaim_usedAsLiteralKey" );
    }

    @Test
    public void testResolve_unmappedValues_dropped()
    {
        run( "testResolve_unmappedValues_dropped" );
    }

    @Test
    public void testResolve_deduplicates()
    {
        run( "testResolve_deduplicates" );
    }

    @Test
    public void testResolve_nonArrayObject_returnsEmpty()
    {
        run( "testResolve_nonArrayObject_returnsEmpty" );
    }

    @Test
    public void testApply_addMode_addsMissingMemberships()
    {
        run( "testApply_addMode_addsMissingMemberships" );
    }

    @Test
    public void testApply_addMode_neverRevokes()
    {
        run( "testApply_addMode_neverRevokes" );
    }

    @Test
    public void testApply_syncMode_revokesMappedNotDesired()
    {
        run( "testApply_syncMode_revokesMappedNotDesired" );
    }

    @Test
    public void testApply_syncMode_leavesNonMappedUntouched()
    {
        run( "testApply_syncMode_leavesNonMappedUntouched" );
    }

    @Test
    public void testApply_missingGroup_skipsWithWarning()
    {
        run( "testApply_missingGroup_skipsWithWarning" );
    }

    @Test
    public void testApply_addMembersThrows_continuesOtherGroups()
    {
        run( "testApply_addMembersThrows_continuesOtherGroups" );
    }

    @Test
    public void testEnsure_createsMissingMappedGroups()
    {
        run( "testEnsure_createsMissingMappedGroups" );
    }

    @Test
    public void testEnsure_idempotent_skipsExisting()
    {
        run( "testEnsure_idempotent_skipsExisting" );
    }

    @Test
    public void testEnsure_deduplicatesMappedGroups()
    {
        run( "testEnsure_deduplicatesMappedGroups" );
    }

    @Test
    public void testEnsure_nullGroups_noop()
    {
        run( "testEnsure_nullGroups_noop" );
    }

    @Test
    public void testEnsure_createThrows_continuesOtherGroups()
    {
        run( "testEnsure_createThrows_continuesOtherGroups" );
    }
}

File: src/test/resources/lib/groupSync-test.js
const test = require('/lib/xp/testing');
const testUtils = require('/testUtils');

const calls = {
    created: [],
    added: [],
    removed: [],
};
let existingGroups = {};
let memberships = [];

const defaultCreateGroup = (params) => {
    calls.created.push(params);
    const key = `group:${params.idProvider}:${params.name}`;
    existingGroups[key] = { key: key };
    return existingGroups[key];
};
const defaultAddMembers = (groupKey, members) => {
    calls.added.push({ group: groupKey, members: members });
};
const defaultRemoveMembers = (groupKey, members) => {
    calls.removed.push({ group: groupKey, members: members });
};

const setAuthMocks = testUtils.mockAndGetUpdaterFunc('/lib/xp/auth', {
    getPrincipal: (key) => existingGroups[key] || null,
    createGroup: defaultCreateGroup,
    addMembers: defaultAddMembers,
    removeMembers: defaultRemoveMembers,
    getMemberships: () => memberships.map(key => ({ key: key })),
});

function resetAuth(opts) {
    opts = opts || {};
    calls.created = [];
    calls.added = [];
    calls.removed = [];
    existingGroups = {};
    (opts.existing || []).forEach(key => { existingGroups[key] = { key: key }; });
    memberships = opts.memberships || [];
    setAuthMocks({
        getPrincipal: (key) => existingGroups[key] || null,
        createGroup: opts.createGroup || defaultCreateGroup,
        addMembers: opts.addMembers || defaultAddMembers,
        removeMembers: opts.removeMembers || defaultRemoveMembers,
        getMemberships: () => memberships.map(key => ({ key: key })),
    });
}

const groupSync = require('./groupSync');

function oktaConfig(extra) {
    const groups = {
        claim: 'groups',
        syncMode: 'add',
        createGroups: true,
        mapping: [
            { value: 'Admins', group: 'group:okta:admins' },
            { value: 'QA-Team', group: 'group:okta:qa' },
            { value: 'Developers', group: 'group:okta:devs' },
        ],
    };
    Object.keys(extra || {}).forEach(k => { groups[k] = extra[k]; });
    return { _idProviderName: 'okta', groups: groups };
}

//-------------------------------------------------------
// resolveGroupKeysFromClaims
//-------------------------------------------------------

exports.testResolve_disabledFeature_returnsEmpty = () => {
    const config = { _idProviderName: 'okta', groups: null };
    test.assertJsonEquals([], groupSync.resolveGroupKeysFromClaims(config, { userinfo: { groups: ['Admins'] } }));
};

exports.testResolve_missingPath_returnsEmpty = () => {
    test.assertJsonEquals([], groupSync.resolveGroupKeysFromClaims(oktaConfig(), { userinfo: { other: ['Admins'] } }));
};

exports.testResolve_topLevelArray_returnsMappedKeys = () => {
    const result = groupSync.resolveGroupKeysFromClaims(oktaConfig(), { userinfo: { groups: ['Admins', 'Developers'] } });
    test.assertJsonEquals(['group:okta:admins', 'group:okta:devs'], result);
};

exports.testResolve_singleString_treatedAsSingleton = () => {
    const result = groupSync.resolveGroupKeysFromClaims(oktaConfig(), { userinfo: { groups: 'Admins' } });
    test.assertJsonEquals(['group:okta:admins'], result);
};

exports.testResolve_nestedDottedPath = () => {
    const config = {
        _idProviderName: 'keycloak',
        groups: {
            claim: 'realm_access.roles',
            syncMode: 'add',
            createGroups: true,
            mapping: [{ value: 'admin', group: 'group:keycloak:admins' }],
        },
    };
    const result = groupSync.resolveGroupKeysFromClaims(config, { userinfo: { realm_access: { roles: ['admin', 'other'] } } });
    test.assertJsonEquals(['group:keycloak:admins'], result);
};

exports.testResolve_uriClaim_usedAsLiteralKey = () => {
    const config = {
        _idProviderName: 'auth0',
        groups: {
            claim: 'https://app.example.com/groups',
            syncMode: 'add',
            createGroups: true,
            mapping: [{ value: 'admins', group: 'group:auth0:admins' }],
        },
    };
    const result = groupSync.resolveGroupKeysFromClaims(config, { userinfo: { 'https://app.example.com/groups': ['admins'] } });
    test.assertJsonEquals(['group:auth0:admins'], result);
};

exports.testResolve_urnClaim_usedAsLiteralKey = () => {
    const config = {
        _idProviderName: 'auth0',
        groups: {
            claim: 'urn:example:groups',
            syncMode: 'add',
            createGroups: true,
            mapping: [{ value: 'admins', group: 'group:auth0:admins' }],
        },
    };
    const result = groupSync.resolveGroupKeysFromClaims(config, { userinfo: { 'urn:example:groups': ['admins'] } });
    test.assertJsonEquals(['group:auth0:admins'], result);
};

exports.testResolve_unmappedValues_dropped = () => {
    const result = groupSync.resolveGroupKeysFromClaims(oktaConfig(), { userinfo: { groups: ['Unknown', 'Admins'] } });
    test.assertJsonEquals(['group:okta:admins'], result);
};

exports.testResolve_deduplicates = () => {
    const result = groupSync.resolveGroupKeysFromClaims(oktaConfig(), { userinfo: { groups: ['Admins', 'Admins'] } });
    test.assertJsonEquals(['group:okta:admins'], result);
};

exports.testResolve_nonArrayObject_returnsEmpty = () => {
    // Entra overage indirection shape; should be silently ignored.
    const result = groupSync.resolveGroupKeysFromClaims(oktaConfig(), {
        userinfo: { groups: { _claim_names: { groups: 'src1' }, hasgroups: true } },
    });
    test.assertJsonEquals([], result);
};

//-------------------------------------------------------
// applyGroups
//-------------------------------------------------------

exports.testApply_addMode_addsMissingMemberships = () => {
    resetAuth({ existing: ['group:okta:admins', 'group:okta:devs'] });
    groupSync.applyGroups(oktaConfig(), 'user:okta:bob', ['group:okta:admins', 'group:okta:devs']);

    test.assertEquals(2, calls.added.length);
    test.assertEquals('group:okta:admins', calls.added[0].group);
    test.assertEquals('group:okta:devs', calls.added[1].group);
    test.assertEquals(0, calls.removed.length);
};

exports.testApply_addMode_neverRevokes = () => {
    resetAuth({
        existing: ['group:okta:admins', 'group:okta:qa'],
        memberships: ['group:okta:qa'],
    });
    // 'add' mode with desired=admins only; qa membership must remain untouched.
    groupSync.applyGroups(oktaConfig(), 'user:okta:bob', ['group:okta:admins']);
    test.assertEquals(0, calls.removed.length);
};

exports.testApply_syncMode_revokesMappedNotDesired = () => {
    resetAuth({
        existing: ['group:okta:admins', 'group:okta:qa'],
        memberships: ['group:okta:qa', 'group:okta:admins'],
    });
    groupSync.applyGroups(oktaConfig({ syncMode: 'sync' }), 'user:okta:bob', ['group:okta:admins']);

    test.assertEquals(1, calls.removed.length);
    test.assertEquals('group:okta:qa', calls.removed[0].group);
};

exports.testApply_syncMode_leavesNonMappedUntouched = () => {
    resetAuth({
        existing: ['group:okta:admins'],
        memberships: ['group:okta:admins', 'group:okta:manual', 'group:system:authenticated'],
    });
    groupSync.applyGroups(oktaConfig({ syncMode: 'sync' }), 'user:okta:bob', ['group:okta:admins']);

    // Only mapped-but-not-desired groups may be revoked; here all desired present, none mapped extra.
    test.assertEquals(0, calls.removed.length);
};

exports.testApply_createGroups_true_createsMissingGroup = () => {
    resetAuth({ existing: [] });
    groupSync.applyGroups(oktaConfig(), 'user:okta:bob', ['group:okta:admins']);

    test.assertEquals(1, calls.created.length);
    test.assertEquals('okta', calls.created[0].idProvider);
    test.assertEquals('admins', calls.created[0].name);
    test.assertEquals('admins', calls.created[0].displayName);
    test.assertEquals(1, calls.added.length);
};

exports.testApply_createGroups_false_skipsWithWarning = () => {
    resetAuth({ existing: [] });
    groupSync.applyGroups(oktaConfig({ createGroups: false }), 'user:okta:bob', ['group:okta:admins']);

    test.assertEquals(0, calls.created.length);
    test.assertEquals(0, calls.added.length);
    test.assertEquals(0, calls.removed.length);
};

exports.testApply_addMembersThrows_continuesOtherGroups = () => {
    resetAuth({
        existing: ['group:okta:admins', 'group:okta:devs'],
        addMembers: (groupKey, members) => {
            if (groupKey === 'group:okta:admins') {
                throw 'boom';
            }
            calls.added.push({ group: groupKey, members: members });
        },
    });
    groupSync.applyGroups(oktaConfig(), 'user:okta:bob', ['group:okta:admins', 'group:okta:devs']);

    // admins threw, devs still processed.
    test.assertEquals(1, calls.added.length);
    test.assertEquals('group:okta:devs', calls.added[0].group);
};

exports.testApply_createGroupThrows_continuesOtherGroups = () => {
    resetAuth({
        existing: ['group:okta:devs'],
        createGroup: (params) => {
            if (params.name === 'admins') {
                throw 'cannot create';
            }
            calls.created.push(params);
            return { key: `group:${params.idProvider}:${params.name}` };
        },
    });
    groupSync.applyGroups(oktaConfig(), 'user:okta:bob', ['group:okta:admins', 'group:okta:devs']);

    // admins create threw; devs already existed and was added.
    test.assertEquals(1, calls.added.length);
    test.assertEquals('group:okta:devs', calls.added[0].group);
};